File: src/test/java/com/gitlab/lae/intellij/actions/tree/app/RootActionTest.java
package com.gitlab.lae.intellij.actions.tree.app;

import com.gitlab.lae.intellij.actions.tree.ActionNode;
import com.gitlab.lae.intellij.actions.tree.When;
import com.intellij.ide.DataManager;
import com.intellij.ide.IdePopupManager;
import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.openapi.util.Pair;
import com.intellij.openapi.wm.IdeFocusManager;
import org.junit.Test;

import javax.swing.*;
import java.util.List;

import static com.intellij.openapi.actionSystem.IdeActions.*;
import static java.util.Arrays.asList;
import static java.util.Collections.emptyList;
import static java.util.Collections.singletonList;
import static javax.swing.KeyStroke.getKeyStroke;
import static org.junit.Assert.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public final class RootActionTest {

    private static final class EnableAction extends AnAction {
        boolean enabled;

        @Override
        public void update(AnActionEvent e) {
            super.update(e);
            e.getPresentation().setEnabled(enabled);
        }

        @Override
        public void actionPerformed(AnActionEvent e) {
        }
    }

    @Test
    public void disablesPresentationIfNoSuitableActionFound() {
        var enable = new EnableAction();
        var when = mock(When.class);
        var action = new RootAction(
                "id",
                emptyList(),
                singletonList(Pair.create(enable, when))
        );
        var presentation = new Presentation();
        presentation.setEnabled(true);
        var event = new AnActionEvent(
                null,
                mock(DataContext.class),
                "",
                presentation,
                mock(ActionManager.class),
                0
        );

        assertTrue(presentation.isEnabled());

        enable.enabled = true;
        when(when.test(any())).thenReturn(false);
        action.update(event);
        assertFalse(presentation.isEnabled());

        when(when.test(any())).thenReturn(true);
        action.update(event);
        assertTrue(presentation.isEnabled());

        enable.enabled = false;
        action.update(event);
        assertFalse(presentation.isEnabled());
    }

    @Test
    public void mergingMaintainsCustomActionGroupsIds() {
        var id = "my-custom-group-id";
        var actual = RootAction.merge(
                singletonList(newActionNode(
                        id,
                        When.ALWAYS,
                        singletonList(getKeyStroke("X")),
                        singletonList(newActionNode(
                                "bob",
                                When.ALWAYS,
                                emptyList(),
                                emptyList()
                        ))
                )),
                mock(ActionManager.class),
                mock(IdeFocusManager.class),
                new IdePopupManager(),
                mock(JBPopupFactory.class),
                mock(DataManager.class)
        );
        assertEquals(1, actual.size());
        assertEquals(id, actual.get(0).getId());
    }

    @Test
    public void mergesRootActions() {
        AnAction cut = new EmptyAction("cut", null, null);
        AnAction copy = new EmptyAction("copy", null, null);
        AnAction paste = new EmptyAction("paste", null, null);
        var actionManager = mock(ActionManager.class);
        when(actionManager.getAction(ACTION_CUT)).thenReturn(cut);
        when(actionManager.getAction(ACTION_COPY)).thenReturn(copy);
        when(actionManager.getAction(ACTION_PASTE)).thenReturn(paste);

        var actual = RootAction.merge(
                asList(
                        newActionNode(
                                ACTION_CUT,
                                When.ALWAYS,
                                singletonList(getKeyStroke('a')),
                                emptyList()
                        ),
                        newActionNode(
                                ACTION_COPY,
                                When.toolWindowActive("Project"),
                                asList(getKeyStroke('a'), getKeyStroke('b')),
                                emptyList()
                        ),
                        newActionNode(
                                ACTION_PASTE,
                                When.fileExtension("txt"),
                                asList(getKeyStroke('x'), getKeyStroke('y')),
                                emptyList()
                        )
                ),
                actionManager,
                null,
                null,
                null,
                null
        );

        var expected = asList(
                new RootAction(
                        "ActionsTree.Root.0",
                        singletonList(getKeyStroke('a')),
                        asList(
                                Pair.create(
                                        copy,
                                        When.toolWindowActive("Project")
                                ),
                                Pair.create(cut, When.ALWAYS)
                        )
                ),
                new RootAction(
                        "ActionsTree.Root.1",
                        singletonList(getKeyStroke('b')),
                        singletonList(Pair.create(
                                copy,
                                When.toolWindowActive("Project")
                        ))
                ),
                new RootAction(
                        "ActionsTree.Root.2",
                        asList(getKeyStroke('x'), getKeyStroke('y')),
                        singletonList(Pair.create(
                                paste,
                                When.fileExtension("txt")
                        ))
                )
        );

        assertEquals(expected, actual);
    }

    private static ActionNode newActionNode(
            String id,
            When when,
            List<KeyStroke> keyStrokes,
            List<ActionNode> items
    ) {
        return ActionNode.create(
                id,
                null,
                null,
                false,
                when,
                keyStrokes,
                items
        );
    }

    private static final class ModalAction extends AnAction {
        ModalAction(boolean enableInModalContext) {
            setEnabledInModalContext(enableInModalContext);
        }

        @Override
        public void actionPerformed(AnActionEvent e) {
        }
    }

    @Test
    public void enableInModalIfAnyActionSupportsModal() {
        assertTrue(new RootAction("id", emptyList(), asList(
                Pair.create(new ModalAction(true), When.NEVER),
                Pair.create(new ModalAction(false), When.NEVER)
        )).isEnabledInModalContext());

        assertFalse(new RootAction("id", emptyList(), asList(
                Pair.create(new ModalAction(false), When.NEVER),
                Pair.create(new ModalAction(false), When.NEVER)
        )).isEnabledInModalContext());
    }
}

File: src/test/kotlin/com/gitlab/lae/intellij/actions/tree/app/RootActionTest.kt
package com.gitlab.lae.intellij.actions.tree.app

import com.gitlab.lae.intellij.actions.tree.When
import com.gitlab.lae.intellij.actions.tree.When.FileExtension
import com.gitlab.lae.intellij.actions.tree.When.ToolWindowActive
import com.gitlab.lae.intellij.actions.tree.actionEvent
import com.gitlab.lae.intellij.actions.tree.actionNode
import com.gitlab.lae.intellij.actions.tree.keys
import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.EmptyAction
import com.intellij.openapi.actionSystem.IdeActions.ACTION_COPY
import com.intellij.openapi.actionSystem.IdeActions.ACTION_CUT
import com.intellij.openapi.actionSystem.IdeActions.ACTION_PASTE
import com.intellij.openapi.actionSystem.Presentation
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

class RootActionTest {

  private class EnableAction : AnAction() {
    var enabled = false

    override fun actionPerformed(e: AnActionEvent) {}

    override fun update(e: AnActionEvent) {
      super.update(e)
      e.presentation.isEnabled = enabled
    }
  }

  @Test
  fun `disables presentation if no suitable action found`() {
    val enable = EnableAction()
    val condition = mock<When>()
    val actionManager = mock<ActionManager>()
    whenever(actionManager.getAction("bob")).thenReturn(enable)

    val action =
        RootAction.merge(
                listOf(actionNode("bob", condition = condition)),
                actionManager,
            )[0]

    val presentation = Presentation()
    presentation.isEnabled = true
    assertTrue(presentation.isEnabled)

    enable.enabled = true
    whenever(condition.test(any())).thenReturn(false)

    val event =
        actionEvent(
            presentation = presentation,
            actionManager = actionManager,
        )
    action.update(event)
    assertFalse(presentation.isEnabled)

    whenever(condition.test(any())).thenReturn(true)
    action.update(event)
    assertTrue(presentation.isEnabled)

    enable.enabled = false
    action.update(event)
    assertFalse(presentation.isEnabled)
  }

  @Test
  fun `merging maintains custom action groups ids`() {
    val id = "my-custom-group-id"
    val actual =
        RootAction.merge(
            listOf(
                actionNode(
                    id = id,
                    keys = keys("X"),
                    items = listOf(actionNode("bob")),
                ),
            ),
            mock(),
        )
    assertEquals(1, actual.size.toLong())
    assertEquals(id, actual[0].id)
  }

  @Test
  fun `merges root actions`() {
    val cut = EmptyAction("Cut", null, null)
    val copy = EmptyAction("Copy", null, null)
    val paste = EmptyAction("Paste", null, null)

    val actionManager = mock<ActionManager>()
    whenever(actionManager.getAction(ACTION_CUT)).thenReturn(cut)
    whenever(actionManager.getAction(ACTION_COPY)).thenReturn(copy)
    whenever(actionManager.getAction(ACTION_PASTE)).thenReturn(paste)

    val cutNode = actionNode(ACTION_CUT, keys = keys("typed a"))
    val copyNode =
        actionNode(
            ACTION_COPY,
            keys = keys("typed a", "typed b"),
            condition = ToolWindowActive("Project"),
        )
    val pasteNode =
        actionNode(
            ACTION_PASTE,
            keys = keys("typed x", "typed y"),
            condition = FileExtension("txt"),
        )

    val actual = RootAction.merge(listOf(cutNode, copyNode, pasteNode), actionManager)

    val expected =
        listOf(
            RootAction(
                "ActionsTree.Root.0",
                keys("typed a"),
                listOf(copyNode, cutNode),
            ),
            RootAction(
                "ActionsTree.Root.1",
                keys("typed b"),
                listOf(copyNode),
            ),
            RootAction(
                "ActionsTree.Root.2",
                keys("typed x", "typed y"),
                listOf(pasteNode),
            ),
        )
    assertEquals(expected, actual)
  }

  @Test
  fun `merging maintains action with no key strokes`() {
    val cut = EmptyAction("Cut", null, null)
    val copy = EmptyAction("Copy", null, null)
    val paste = EmptyAction("Paste", null, null)
    val actionManager = mock<ActionManager>()
    whenever(actionManager.getAction(ACTION_CUT)).thenReturn(cut)
    whenever(actionManager.getAction(ACTION_COPY)).thenReturn(copy)
    whenever(actionManager.getAction(ACTION_PASTE)).thenReturn(paste)

    val cutNode = actionNode(ACTION_CUT)
    val copyNode = actionNode(ACTION_COPY, keys = keys("typed b"))
    val pasteNode = actionNode(ACTION_PASTE)

    val actual = RootAction.merge(listOf(cutNode, copyNode, pasteNode), actionManager)

    val expected =
        listOf(
            RootAction(
                "ActionsTree.Root.0",
                keys("typed b"),
                listOf(copyNode),
            ),
            RootAction(
                "ActionsTree.Root.1",
                emptyList(),
                listOf(pasteNode, cutNode),
            ),
        )
    assertEquals(expected, actual)
  }

  private class ModalAction(enableInModalContext: Boolean) : AnAction() {
    override fun actionPerformed(e: AnActionEvent) {}

    init {
      isEnabledInModalContext = enableInModalContext
    }
  }

  @Test
  fun `enable in modal if any action supports modal`() {
    val modalTrue = ModalAction(true)
    val modalFalse = ModalAction(false)

    val actionManager = mock<ActionManager>()
    whenever(actionManager.getAction("true")).thenReturn(modalTrue)
    whenever(actionManager.getAction("false")).thenReturn(modalFalse)

    assertTrue(
        RootAction.merge(
                listOf(
                    actionNode("true"),
                    actionNode("false"),
                ),
                actionManager,
            )[0]
            .isEnabledInModalContext,
    )

    assertFalse(
        RootAction.merge(
                listOf(
                    actionNode("false"),
                    actionNode("false"),
                ),
                actionManager,
            )[0]
            .isEnabledInModalContext,
    )
  }
}
